docs(*): add docstrings - #37

K-Kumar-01 opened the pull request:
<!--- Provide a formatted commit message describing this PR in the Title above -->
<!--- See our DEVELOPERS guide below: -->
<!--- https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format -->
# Closes #37 <CORRESPONDING ISSUE NUMBER>
<!--- Provide an overall summary of the pull request -->
Added the docstrings for the project.

### Changes
<!--- More detailed and granular description of changes -->
<!--- These should likely be gathered from commit message summaries -->
- <ONE> Added the package `eslint-plugin-jsdoc` and configured `.eslintrc.json`.
- <TWO> Added docstrings for the various functions in the project.

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [x] Extend the documentation, if necessary
- [x] Merging to `master` from `fork:branchname`
- [ ] Manual accessibility test performed
    - [ ] Keyboard-only access, including forms
    - [ ] Contrast at least WCAG Level A
    - [ ] Appropriate labels, alt text, and instructions

DianaLease commented on src/NewSmartClauseDialog.js:
```diff
@@ -42,17 +42,17 @@ class NewSmartClauseDialog extends Component {
 
   handleCancel = () => {
-    this.setState({ open: false });
+    this.setState({ open: false,clauseId: '', templateId: '' });
```
super small change, but can you add a space before `clauseId` please ?

shakti97 replied:
Done

shakti97 commented:
@DianaLease can you please review and suggest changes if any

K-Kumar-01 committed:
chore: Eslint setup and configuration for js-docs - #37

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

K-Kumar-01 committed:
docs: Added docstring for functions - #37

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

algomaster99 commented on package.json:
```diff
@@ -49,6 +49,7 @@
     "copy-webpack-plugin": "^6.0.1",
     "css-loader": "^3.0.0",
     "eslint": "^7.1.0",
+    "eslint-plugin-jsdoc": "^32.3.0",
```
Even after adding this package, I don't get an error when I run `npm run lint` even though I have removed all docstrings from my local.

K-Kumar-01 replied:
@algomaster99 
This package does not enforce js-doc to be present.
I found that we could go with the package `eslint-plugin-require-jsdoc`.
However, this rule was deprecated as mentioned [here](https://eslint.org/docs/rules/require-jsdoc).
So should we go with this one? 🤔

algomaster99 replied:
> This package does not enforce js-doc to be present.

Why have you added it then? Can you tell its purpose? Does it only check the syntax of the JSDoc?

K-Kumar-01 replied:
@algomaster99 
The purpose of the above-added package was to enforce valid js-docs. It ensures that JSDoc follows certain rules. More details [here](https://www.npmjs.com/package/eslint-plugin-jsdoc). To enforce that JSDoc is present `eslint-plugin-require-jsdoc` should be installed (deprecated) as I mentioned [here](https://github.com/accordproject/cicero-word-add-in/pull/38#discussion_r606952209).

K-Kumar-01 replied:
@algomaster99 
I was able to figure that the eslint rules for the arrow function and other things were false by default, which resulted in no warning being thrown for missing docs.
Hence we do no require `eslint-plugin-require-jsdoc`. I will update my PR with the new rules ASAP.

algomaster99 commented on src/components/App.js:
```diff
@@ -20,6 +20,11 @@ const App = ({ isOfficeInitialized }) => {
   };
 
+  /**
+   * Change the startup state of the add in
```
Add a full stop

algomaster99 replied:
Be more descriptive.

`Set whether add-in should load by default on MS Word startup.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -19,6 +19,9 @@ const LibraryComponent = () => {
 
   useEffect(() => {
+    /**
+     * Loading the template library and storing them in the state.
```
Mention the URL too. [https://templates.accordproject.org/](https://templates.accordproject.org/).

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -30,6 +33,11 @@ const LibraryComponent = () => {
   }, []);
 
+  /**
+   * Upload a template locally
```
If you're adding a full stop in the **main description** (not the blocks), make sure you do it everywhere.

algomaster99 replied:
Better description would be `Render the uploaded template.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -30,6 +33,11 @@ const LibraryComponent = () => {
+  /**
+   * Upload a template locally
+   *
+   * @param {MouseEvent} event File upload
```
`event containing the file object`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -43,6 +51,9 @@ const LibraryComponent = () => {
 
   useEffect(() => {
+    /**
+     * Initializing the document for the particular template with necssary uitlities whenver templates state changes
```
The spellings of "necessary", "utilities", and "whenever" are wrong. Please correct it.

algomaster99 replied:
`Initialize the document by fetching the templates whose identifier is stored in CustomXMLPart.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -79,6 +90,13 @@ const LibraryComponent = () => {
 
+
+  /**
+   * Does the initial setup when the template is inserted.
```
`Sets up a template for rendering.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -130,6 +148,12 @@ const LibraryComponent = () => {
   };
 
+  /**
+   * Converts a template to ciceromark
```
`Converts a template text to CiceroMark JSON.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -147,6 +176,11 @@ const LibraryComponent = () => {
   };
 
+  /**
+   * Save the template details to xml
```
`Save the template details to CustomXML.`

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -138,6 +162,11 @@ const LibraryComponent = () => {
   };
 
+  /**
+   * Loads the template text, setup it and save the template to xml
```
`Fetches templateIndex from https://templates.accordproject.org/, load the template, and save template details to CustomXML. `

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -186,6 +220,11 @@ const LibraryComponent = () => {
   };
 
+  /**
+   * Redirect to the template url
```
URL

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -186,6 +220,11 @@ const LibraryComponent = () => {
+  /**
+   * Redirect to the template url
+   *
+   * @param {object} template Template Object
```
`object`

algomaster99 commented on src/index.js:
```diff
@@ -4,7 +4,11 @@ import * as React from 'react';
+ *
+ * @param {React.ReactDOM} Component
+ * Renders a react component
+ */
```
```js
/**
 * Renders a react component.
 * 
 * @param {React.Component} React component
 */
```

algomaster99 commented on src/utils/AttachVariableChangeListener.js:
```diff
@@ -1,3 +1,8 @@
+ *
+ * @param {string} title
+ * Title of the variable
+ */
```
Attaches listener to content controls so that similar instances of variables are updated simultaneously.

algomaster99 commented on src/utils/AttachVariableChangeListener.js:
```diff
@@ -18,6 +23,13 @@ const attachVariableChangeListener = title => {
 
+
+/**
+ * Change variables with same tag if one of the variable changes among them i.e. make all variables with same tag have similar value
```
Well written! :) Only grammatical fix.

`Updates value of all the instance of variables with the same tags when one of them is manually modified.`

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -2,7 +2,16 @@ import sanitizeHtmlChars from './SanitizeHtmlChars';
+ * Transforms the given heading node into OOXML heading
+ *
+ * @param {string} value
+ * Text to be rendered as heading
```
Write it beside `value`.

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -26,10 +35,20 @@ const insertHeading = (value, level) => {
 };
 
+/**
+ * Insert a line break
```
Inserts a line break.

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -26,10 +35,20 @@ const insertHeading = (value, level) => {
 };
 
+/**
+ * Insert a soft break
```
Follow the same tense like I have done above.

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -82,6 +117,13 @@ const insertVariable = (title, tag, value, type) => {
+ * Inserts list
+ *
+ * @param {Array} node Array of nodes
+ * @param {string} type Type of list- ordered or unordered
```
Type of list - ordered or unordered (Notice the space)

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -237,6 +294,14 @@ const getNodes = (node, counter, parent=null) => {
+ * @param {Array} ciceroMark Ciceromark JSON
+ * @param {object} counter Counter for different variables based on node name
+ * @param {string} ooxml intial OOXML string
+ * @returns {string} converted OOXML string i.e. Cicecomark->OOXML
```
CiceroMark (check your spelling)

algomaster99 reviewed (changes requested):
Nice work is done! A few changes and then we might be good to go.

I suggest you follow the standards here https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/.

K-Kumar-01 committed:
docs: Updated the docstrings - #37

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

K-Kumar-01 commented:
@algomaster99 
I have updated the docstrings. Let me know if there is any change required.

Also when I run `npm run lint` i get errors in my console.
![Screenshot (38)](https://user-images.githubusercontent.com/59891164/113547471-9bf0fa80-960b-11eb-9dff-c21bdb6c9908.png)
I got [this ](https://github.com/eslint/eslint/issues/10880). I also changed the `package.json` script from `"lint": "eslint **/*.{js,jsx}",` to `"lint": "eslint '**/*.{js,jsx}'",`
Also there was a change in file to check if eslint worked
![Screenshot (39)](https://user-images.githubusercontent.com/59891164/113547935-5b45b100-960c-11eb-8ba4-aefbaa2297dd.png)
However I did not get any error messages in the console. Could you please explain it what I am doing wrong?

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -30,6 +33,11 @@ const LibraryComponent = () => {
   }, []);
 
+  /**
+   * Render a uploaded template.
```
"an uploaded"

algomaster99 commented on src/components/TemplateLibrary/index.js:
```diff
@@ -19,6 +19,9 @@ const LibraryComponent = () => {
 
   useEffect(() => {
+    /**
+     * Loading the template library from https://templates.accordproject.org/ and storing them in the state.
```
Loads.

Use third-person verb form everywhere.

K-Kumar-01 committed:
chore: Eslint setup and configuration for js-docs - #37

Updated rules
Updated scripts

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

K-Kumar-01 committed:
docs: Updated the docstring - #37

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

K-Kumar-01 commented:
@algomaster99 @irmerk 
I have updated the PR.
Now, it will warn if js-docs are missing from the project.
Also, the `npm run lint` script has been fixed. 
Let me know if there are any changes required.

algomaster99 commented on package.json:
```diff
@@ -16,8 +16,8 @@
-    "lint": "eslint **/*.{js,jsx}",
-    "lint:fix": "eslint **/*.{js,jsx} --fix",
+    "lint": "eslint '**/*.{js,jsx}'",
+    "lint:fix": "eslint '**/*.{js,jsx}' --fix",
```
Why are these quotes needed?

K-Kumar-01 replied:
@algomaster99 
If no quotation marks are applied, then [this](https://github.com/eslint/eslint/issues/10880) issue occurs.
So the quotation marks are required for the lint command to run properly.

algomaster99 commented on src/utils/VariableVisitor.js:
```diff
@@ -1,18 +1,44 @@
   }
 
+  /**
+   * Update the counters for variable fields and visit the sub fields for other types.
```
Write the verbs in third-person form as the majority of the docstrings you have written are in the same form.

algomaster99 commented on src/utils/CiceroMarkToOOXML.js:
```diff
@@ -154,6 +201,14 @@ const definedNodes = {
 };
 
+/**
+ * Get the OOXMl for the given node.
```
OOXM**L**

Notice the capital L.

algomaster99 commented on src/error/BadFile/index.js:
```diff
@@ -3,7 +3,11 @@ import * as ReactDOM from 'react-dom';
 import './index.css';
-
+/**
+ * Renders a message when bad file is uploaded
```
Please add a full-stop at the end.

algomaster99 reviewed (changes requested):
Some changes left and we are good to go.

I haven't requested the change only on the lines I have dropped my review. Just check yourself wherever you're inconsistent and fix that. I don't want you to stick the rules. I just want you to be consistent.

K-Kumar-01 committed:
docs: Updated the docstrings - #37

Usage of third-person verb form to maintain consistency

Signed-off-by: K-Kumar-01 <59891164+K-Kumar-01@users.noreply.github.com>

K-Kumar-01 commented:
> Some changes left and we are good to go.
> 
> I haven't requested the change only on the lines I have dropped my review. Just check yourself wherever you're inconsistent and fix that. I don't want you to stick the rules. I just want you to be consistent.

@algomaster99 
I have updated the PR to use the third-person verb forms to maintain consistency. Also, full stops are added to the description of each function
Let me know if there are any changes required.

algomaster99 reviewed (approved):
I think this is good to be merged now :)
Nice work :tada:
